nup/update.go

```diff
@@ -0,0 +1,175 @@
+// is only meaningful if the operation is OpSet.
+func (u Update[T]) ValueOperation() (value T, operation Operation) {
+	return u.value, u.op
+}
+
+// Operation returns the operation this update performs: no-op, remove, or set.
+func (u Update[T]) Operation() Operation {
+	return u.op
+}
+
+// IsNoop returns whether this update is a no-op. IsNoop is equivalent to
+// Operation() == OpNoop.
+func (u Update[T]) IsNoop() bool {
+	return u.op == OpNoop
+}
+
+// IsRemove returns whether this update is a remove operation. IsRemove is
+// equivalent to Operation() == OpRemove.
+func (u Update[T]) IsRemove() bool {
+	return u.op == OpRemove
+}
+
+// IsSet returns whether this update is a set operation. IsSet is equivalent to
+// Operation() == OpSet.
+func (u Update[T]) IsSet() bool {
+	return u.op == OpSet
+}
+
+// IsChange returns whether this update is either a set or remove operation
+// (i.e., not a no-op). IsChange is equivalent to Operation() != OpNoop.
+func (u Update[T]) IsChange() bool {
+	return u.op != OpNoop
+}
+
+// Value returns the value this update sets fields to (if any) and an isSet flag
+// indicating whether the update is a set operation. If the flag is false
+// (because the update is actually a no-op or removal), then the returned value
+// is T's zero value.
+func (u Update[T]) Value() (value T, isSet bool) {
+	return u.value, u.op == OpSet
+}
+
+// ValueOrNil returns this update's value if it's a set operation or else nil.
+func (u Update[T]) ValueOrNil() *T {
+	if u.op != OpSet {
+		return nil
+	}
+	// Copy the update value so it can't be mutated via the returned pointer.
+	value := u.value
+	return &value
+}
+
+// Apply returns the result of applying the update to the given value. The
+// result is the given value if the update is a no-op, the zero value if it's a
+// removal, or the update's contained value is if it's a set operation.
+func (u Update[T]) Apply(value T) T {
+	switch u.op {
+	case OpNoop:
+		return value
+	case OpRemove:
+		var zero T
+		return zero
+	default: // Set
+		return u.value
+	}
+}
+
+// Diff returns the update itself if Apply(value) != value; otherwise it returns
+// a no-op update. Diff can be used to omit extraneous updates when applying
+// them would have no effect.
+func (u Update[T]) Diff(value T) Update[T] {
+	if u.Apply(value) == value {
+		return Noop[T]()
+	}
+	return u
+}
+
+// UnmarshalJSON implements json.Unmarshaler.
+func (u *Update[T]) UnmarshalJSON(data []byte) error {
+	if string(data) == "null" {
+		u.op = OpRemove
+		return nil
+	}
+	u.op = OpSet
+	return json.Unmarshal(data, &u.value)
+}
+
+// IsSetTo returns whether the update sets to the given value.
+func (u Update[T]) IsSetTo(value T) bool {
+	return u.op == OpSet && u.value == value
+}
+
+// IsSetSuchThat returns whether the update is a set operation to a value that
+// satisfies the given predicate.
+func (u Update[T]) IsSetSuchThat(predicate func(T) bool) bool {
+	return u.op == OpSet && predicate(u.value)
+}
```

Comment:
I ended up not using this method anywhere in my `entity` branch, mainly because anonymous functions are pretty bulky in go. I think this still might sometimes be handy for one-off checks though, especially when an existing function can be passed in.